src/lib.rs

```diff
@@ -440,6 +440,68 @@ impl KmerCountTable {
         self.counts.insert(hashval, count);
         Ok(())
     }
+
+    /// Calculates the Jaccard Similarity Coefficient between two KmerCountTable objects.
+    /// # Returns
+    /// The Jaccard Similarity Coefficient between the two tables as a float value between 0 and 1.
+    pub fn jaccard(&self, other: &KmerCountTable) -> f64 {
+        // Get the intersection of the two k-mer sets.
+        let intersection_size = self.intersection(other).len();
+
+        // Get the union of the two k-mer sets.
+        let union_size = self.union(other).len();
+
+        // Handle the case where the union is empty (both sets are empty).
+        if union_size == 0 {
+            return 1.0; // By convention, two empty sets are considered identical.
+        }
+
+        // Calculate and return the Jaccard similarity as a ratio of intersection to union.
+        intersection_size as f64 / union_size as f64
+    }
+
+    /// Cosine similarity between two `KmerCountTable` objects.
+    /// # Returns
+    /// The cosine similarity between the two tables as a float value between 0 and 1.
+    pub fn cosine(&self, other: &KmerCountTable) -> f64 {
+        // Early return if either table is empty.
+        if self.counts.is_empty() || other.counts.is_empty() {
+            return 0.0;
+        }
+
+        // Calculate the dot product in parallel.
+        let dot_product: u64 = self
+            .counts
+            .par_iter()
+            .filter_map(|(&hash, &count1)| {
+                // Only include in the dot product if both tables have the k-mer.
+                other.counts.get(&hash).map(|&count2| count1 * count2)
```

Comment:
I don't immediately understand the error handling here - `get` returns a `None` if `&hash` is not present, but does `.map` operate on a `None`??

Reply:
I don't think there should be any errors to handle, `get()` will return `None` if the query hash is not present in the _other_ table. I think `map()` should skip anything that returns `None` and not attempt to multiply count1 * count2.

Seeing as we sum all the results at the end the missing zeros shouldn't make a difference.